Services/Repository.cs

```diff
@@ -0,0 +1,22 @@
+using System.Linq.Expressions;
+
+
+namespace TodoNET6.Services
+{
+    public interface IEntity<TIndex>
+    {
+        TIndex Id { get; set; }
+    }
+
+    public interface IRepository<TEntity, TIndex>
+        where TEntity : IEntity<TIndex>
+    {
+        Task<TEntity> CreateAsync(TEntity entity);
+        Task<TEntity> FetchAsync(TIndex index);
+        Task<IEnumerable<TEntity>> WhereAsync(
+            Expression<Func<TEntity, bool>> predicate
```

Comment:
What are the `Expression` and `Func` types here? Is that why you're using linq?

Reply:
`Func` is, obviously, just the way you represent a function variable in the type system.

`Expression` is a bit more interesting as an `Expression` represents the model of a literal `C# statement` or similar that is not compiled. This is related to the reflection system and `Expression` is easily one of the most complex parts of .NET.

You can see `Expression` to dynamically generate compiled programs at runtime (with strong typing pre-runtime). In this case I am using `Expression` here because most C# based ORMs take an `Expression` object as a predicate to build queries via. reflection.

Technically here you can just use a `Func` but in EF Core and LINQ in general if you pass a `Func` as an `Expression` then the ORM will have to fetch ALL query results and then apply the predicate in memory rather than using the predicate to construct the query instead.

See: https://stackoverflow.com/questions/64376006/entity-framework-core-not-generating-where-clause-when-using-linq-and-predicate (Ignore the fart-sniffing SO moderators heavily breathing on my post)